File: C++/347_TopKFrequentElements.cpp
using namespace std;
#include <vector>
#include <unordered_map>
#include <priority_queue>

class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {

        unordered_map<int, int> ht;

        for (auto num:nums) {
            ht[num] += 1;
        }

        priority_queue<pair<int,int>, vector<pair<int,int>>, greater<pair<int,int>>> minHeap;

        for (auto& [element, frequency]: ht) {
            minHeap.push({frequency, element});
            if (minHeap.size() > k) minHeap.pop();
        }

        vector<int> ans;

        for (int i=0;i<k;i++) {
            ans.push_back(minHeap.top().second);
            minHeap.pop();
        }

        return ans;        
    }
};

File: Python3/347_TopKFrequentElements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        
        ht = {}

        for num in nums:
            if num in ht:
                ht[num] += 1
            else:
                ht[num] = 1

        hp = []

        for key, value in ht.items():
            heappush(hp, (value, key))
            if len(hp) > k:
                heappop(hp)

        for e in hp:
            print(e)

        ans = []

        for e in hp:
            ans.append(e[1])

        return ans